src/plugins/data_source/server/legacy/configure_legacy_client.ts

```diff
@@ -0,0 +1,167 @@
+    const dataSource = await getDataSource(dataSourceId, savedObjects);
+    const rootClient = getRootClient(dataSource.attributes, config, openSearchClientPoolSetup);
+
+    return await getQueryClient(rootClient, dataSource, cryptographyClient, callApiParams);
+  } catch (error: any) {
+    logger.error(`Fail to get data source client for dataSourceId: [${dataSourceId}]`);
+    logger.error(error);
+    // Re-throw as DataSourceConfigError
+    throw new DataSourceConfigError('Fail to get data source client: ', error);
+  }
+};
+
+/**
+ * With given auth info, wrap the rootClient and return
+ *
+ * @param rootClient root client for the connection with given data source endpoint.
+ * @param dataSource data source saved object
+ * @param cryptographyClient cryptography client for password encryption / decryption
+ * @returns child client.
+ */
+const getQueryClient = async (
+  rootClient: Client,
+  dataSource: SavedObject<DataSourceAttributes>,
+  cryptographyClient: CryptographyClient,
+  { endpoint, clientParams, options }: LegacyClientCallAPIParams
+) => {
+  const authType = dataSource.attributes.auth.type;
+
+  switch (authType) {
+    case AuthType.NoAuth:
+      return await (callAPI.bind(null, rootClient) as LegacyAPICaller)(
+        endpoint,
+        clientParams,
+        options
+      );
+    case AuthType.UsernamePasswordType:
+      const credential = await getCredential(dataSource, cryptographyClient);
+      return getBasicAuthClient(rootClient, { endpoint, clientParams, options }, credential);
+
+    default:
+      throw Error(`${authType} is not a supported auth type for data source`);
+  }
+};
+
+/**
+ * Gets a root client object of the OpenSearch endpoint.
+ * Will attempt to get from cache, if cache miss, create a new one and load into cache.
+ *
+ * @param dataSourceAttr data source saved objects attributes.
+ * @param config data source config
+ * @returns Legacy client for the given data source endpoint.
+ */
+const getRootClient = (
+  dataSourceAttr: DataSourceAttributes,
+  config: DataSourcePluginConfigType,
+  { getClientFromPool, addClientToPool }: OpenSearchClientPoolSetup
+): Client => {
+  const endpoint = dataSourceAttr.endpoint;
+  const cachedClient = getClientFromPool(endpoint);
+  if (cachedClient) {
+    return cachedClient as Client;
+  } else {
+    const configOptions = parseClientOptions(config, endpoint);
+    const client = new Client(configOptions);
+    addClientToPool(endpoint, client);
+
+    return client;
+  }
+};
+
+/**
+ * Calls the OpenSearch API endpoint with the specified parameters.
+ * @param client Raw legacy JS client instance to use.
+ * @param endpoint Name of the API endpoint to call.
+ * @param clientParams Parameters that will be directly passed to the
+ * legacy JS client.
+ * @param options Options that affect the way we call the API and process the result.
+ * make wrap401Errors default to false, because we don't want login pop-up from browser
+ */
+const callAPI = async (
```

Comment:
could we add comment about the reason and source of duplication of this snippet?